### PR TITLE
Resolving ILLink warnings on System.Private.Xml (Part 1)

### PR DESCRIPTION
**Pull request by joperezr**
Contributes to https://github.com/dotnet/runtime/issues/45623

cc: @eerhardt @krwq

This is the first part of resolving the System.Private.Xml linker warnings, which for now involves most of the XslCompiledTransform ones.

### CODE REVIEW

**Comment by joperezr** [comment]
Some of the annotations touch public  API so I'll rerun genapi to have those annotations on the ref as well.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/QilGenerator.cs [hunk 1]
```diff
@@ -196,13 +196,14 @@ private QilExpression Compile(Compiler compiler)
             }
 
             // Create list of all early bound objects
-            Dictionary<string, Type?> scriptClasses = compiler.Scripts.ScriptClasses;
+            Scripts.LinkerSafeDictionary scriptClasses = compiler.Scripts.ScriptClasses;
```
Comment (vitek-karas) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/QilGenerator.cs:199]: Nit: Try to not use the term `link` outside of the linker repo - it's confusing since it typically means different thing (platform linker). We try to stick to the term `trim`/`trimmer` everywhere.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs [hunk 2]
```diff
@@ -41,5 +43,44 @@ public Scripts(Compiler compiler)
             }
             return null;
         }
+
+        internal class LinkerSafeDictionary : IDictionary<string, Type?>
```
Comment (vitek-karas) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs:47]: This works - but do we really need to implement the full `IDictionary`? This class is used in internal code only, so we could probably get by with only the necessary minimum, right?

Ideally I would like to avoid the suppressions for IL2093/IL2092.

File: src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs [hunk 3]
```diff
@@ -89,6 +89,7 @@ private void Reset()
 
         // SxS: This method does not take any resource name and does not expose any resources to the caller.
         // It's OK to suppress the SxS warning.
+        [RequiresUnreferencedCode("This method requires a call to LoadInternal which might use types that cannot be statically analyzed.")]
```
Comment (vitek-karas) [inline src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs:92]: So this means there's no way to use XSLT and be "trim compatible" right?
I haven't looked into the details - maybe that's really the case. In any case I would probably create an issue to track this -  in theory running XSLT should not break trimming (precompiled XSLT is a separate problem) - in most cases.

Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs:92]: Yeah, I wasn't familiar with xslt but have been playing around with the scenarios, and there is actually only one scenario that should be unsafe:
We have a tool called xsltc.exe that takes in an xsl file and compiles it into a generated managed assembly, and we have an entry point from XslCompiledTransform.Load that can take in a pointer to this generated managed assembly and will internally call methods inside that assembly using reflection, which will of course be not trim safe. The reason why I had to annotate all Load methods is because they internally all use a common LoadInternal which I had to annotate as it also has this unsafe scenario included in it, but I will refactor so that only this specific scenario is marked as unsafe but not the rest.

Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs:92]: I'd assume that if you want to use XslCompiledTransform you'd likely need most of the types from Xsl.Runtime namespace.

Comment (vitek-karas) [reply src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs:92]: If there are dependencies those should be either found by the trimmer automatically, or we will need to add trimmer annotations (like `DynamicDependency`). If it ends up including most of `Xsl.Runtime`, so be it. The point of this discussion is that it should be possible to run XSLT even in a trimmed app without issues. Unless it uses precompiled XSLT, that won't work.

Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs:92]: > that it should be possible to run XSLT even in a trimmed app without issues.

I made at least the simplest case work in #45522, and added a trimming test for it.

Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs]: Yeah, if you guys check my latest change on this PR, I've set it so that only two Load overloads are marked as RUC which are the ones that take in a Type and a MethodInfo of an assembly that was generated by xsltc.exe when passed in a xsl. This will of course always be unsafe, but all other Load overloads are fine because even though some of the types they instantiate cannot be statically analyzed, all of these types are generated using ref emit at runtime so the trimmer can't remove those. Most of Xsl.Runtime is preserved, but that is just because the linker can statically analyze that it needs it so no issues there.

**Comment by joperezr** [comment]
This is almost ready for review, but I'm still addressing a couple of warnings still so hold off on reviewing for now

**Comment by joperezr** [comment]
This is now ready for review

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/EarlyBoundInfo.cs [hunk 4]
```diff
@@ -14,13 +15,16 @@ internal sealed class EarlyBoundInfo
     {
         private readonly string _namespaceUri;            // Namespace Uri mapped to these early bound functions
         private readonly ConstructorInfo _constrInfo;     // Constructor for the early bound function object
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
+        private readonly Type _ebType;
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/EarlyBoundInfo.cs:19]: (minor) I assume we don't need to worry about the size of these objects.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/EarlyBoundInfo.cs [hunk 5]
```diff
@@ -33,7 +37,11 @@ public EarlyBoundInfo(string namespaceUri, Type ebType)
         /// <summary>
         /// Return the Clr Type of the early bound object.
         /// </summary>
-        public Type EarlyBoundType { get { return _constrInfo.DeclaringType; } }
+        [property: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/EarlyBoundInfo.cs:40]:
```suggestion
        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
```

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/QilGenerator.cs [hunk 6]
```diff
@@ -196,13 +196,14 @@ private QilExpression Compile(Compiler compiler)
             }
 
             // Create list of all early bound objects
-            Dictionary<string, Type?> scriptClasses = compiler.Scripts.ScriptClasses;
+            Scripts.TrimSafeDictionary scriptClasses = compiler.Scripts.ScriptClasses;
             List<EarlyBoundInfo> ebTypes = new List<EarlyBoundInfo>(scriptClasses.Count);
-            foreach (KeyValuePair<string, Type?> pair in scriptClasses)
+            foreach (string key in scriptClasses.Keys)
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/QilGenerator.cs:201]: Why the change here to enumerate the keys and then index into the dictionary? Why not just expose `GetEnumerator` from `TrimSafeDictionary`?

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs [hunk 7]
```diff
@@ -41,5 +43,29 @@ public Scripts(Compiler compiler)
...
+        internal class TrimSafeDictionary
+        {
+            private readonly Dictionary<string, Type?> _backingDictionary = new Dictionary<string, Type?>();
+
+            public Type? this[string key]
+            {
+                [UnconditionalSuppressMessage("TrimAnalysis", "IL2073:MissingDynamicallyAccessedMembers",
+                    Justification = "The getter of the dictionary is not annotated to preserve the constructor, but the sources that are adding the items to " +
+                    "the dictionary are annotated so we can supress the message as we know the constructor will be preserved.")]
+                [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs:56]: Can we just put the attribute on the whole property?

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs [hunk 8]
```diff
@@ -371,15 +371,10 @@ internal XsltOutput Output
 
             _scriptExtensions = new Hashtable(_stylesheet.ScriptObjectTypes.Count);
             {
-                foreach (DictionaryEntry entry in _stylesheet.ScriptObjectTypes)
+                // Scripts are not supported on stylesheets
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs:374]: Do we need a test for this or something? Are we concerned about the behavior change?

Comment (vitek-karas) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs:374]: This is a functional change as far as I can tell - we should have the proper tests and get it reviewed by the area owners.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/QilGenerator.cs (hunk 6 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/QilGenerator.cs:201]: The problem here is that if we keep it as it was, then we would get a warning (even when using TrimmerSafeDictionary) since the Type would be comming form method: `System.Collections.Generic.KeyValuePair<Tkey, TValue>.get` which wouldn't be annotated. Of course we could suppress that, but by doing it the way I'm doing it here, the value gets accessed via the index operator on TrimmerSafeDictionary, so that way the linker will not throw a warning and we won't need an additional suppression.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs (hunk 8 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs:374]: Agreed, @krwq AFAICT Script objects on xsls are not supported at all in .NET Core. I tried with several different approaches trying to create an xsl that had scripts and all entrypoints that I found trhough XslCompiledTransform would all eventually throw PNSE. I think that this has been dead code that we just originally imported from .NET Framework, but never really did anything as it throws PNSE when you actually invoke the transform operation.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/EarlyBoundInfo.cs (hunk 4 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/EarlyBoundInfo.cs:19]: @krwq do you know if that's the case? Also, do you know if EarlyBoundInfos get used for anything other than Scripts in Xsls? Because if that's the case, we can potentially just remove this whole type as it would be currently just dead code.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs (hunk 8 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs:374]: IIRC that's correct and intentional. Might be a good idea to remove this code completely and just shortcut to PNSE to save some bytes.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs [hunk 9]
```diff
@@ -371,15 +371,10 @@ internal XsltOutput Output
...
+                if (_stylesheet.ScriptObjectTypes.Count > 0)
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs:375]: might be a good idea to also simplify parsing this to save even more bytes

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs (hunk 7 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs:56]: Chatted offline, this is a problem specific to Indexers and have logged the following issue for investigation tracking: https://github.com/mono/linker/issues/1902

File: src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs [hunk 10]
```diff
@@ -238,6 +238,7 @@ public void Load(Type compiledStylesheet)
                 throw new ArgumentException(SR.Format(SR.Xslt_NotCompiledStylesheet, compiledStylesheet.FullName), nameof(compiledStylesheet));
         }
 
+        [RequiresUnreferencedCode("This method will call into constructors of the earlyBoundTypes array which cannot be statically analized.")]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs:241]: `analized` => `analyzed` repeated elsewhere.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XmlIlGenerator.cs [hunk 11]
```diff
@@ -70,6 +70,10 @@ public XmlILGenerator()
         // SxS Note: The way the trace file names are created (hardcoded) is NOT SxS safe. However the files are
         // created only for internal tracing purposes. In addition XmlILTrace class is not compiled into retail
         // builds. As a result it is fine to suppress the FxCop SxS warning.
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:MissingRequiresUnreferencedCode",
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/XmlIlGenerator.cs:73]: (nit) typically I've been using just `IL2026:RequiresUnreferencedCode`. The word after the number doesn't really matter to the trimmer. But warning IL2026 is because you are calling a method attributed with `RequiresUnreferencedCode`. So when you suppress the warning, you are suppressing the fact that you are calling a method with RequiresUnreferencedCode, not that RequiresUnreferencedCode is "missing".

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XmlIlGenerator.cs [hunk 12]
```diff
@@ -70,6 +70,10 @@ public XmlILGenerator()
...
+            Justification = "This method will generate the IL methods using RefEmit at runtime, which will then try to call them " +
+            "using methods that are annotated as RequiresUnreferencedCode. In this case, these uses can be suppressed as the " +
+            "linker won't be able to trim any IL that gets generated at runtime.")]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/XmlIlGenerator.cs:76]: `linker` => `trimmer`

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs [hunk 13]
```diff
@@ -371,15 +371,10 @@ internal XsltOutput Output
...
                 {
-                    string namespaceUri = (string)entry.Key;
-                    if (GetExtensionObject(namespaceUri) != null)
-                    {
-                        throw XsltException.Create(SR.Xslt_ScriptDub, namespaceUri);
-                    }
-                    _scriptExtensions.Add(namespaceUri, Activator.CreateInstance((Type)entry.Value!,
-                        BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, null, null));
+                    throw new PlatformNotSupportedException(SR.CompilingScriptsNotSupported);
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs:377]: Is it worth/possible to add a test here? I'm not asking for a bunch of work, but if it is reasonable (say an hour) it might be nice to ensure this scenario throws PNSE correctly.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs [hunk 14]
```diff
@@ -41,5 +43,29 @@ public Scripts(Compiler compiler)
...
+                get => ((IDictionary<string, Type?>)_backingDictionary)[key];
+                [param: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
+                set => ((IDictionary<string, Type?>)_backingDictionary)[key] = value;
+            }
+
+            public ICollection<string> Keys => ((IDictionary<string, Type?>)_backingDictionary).Keys;
+
+            public int Count => ((ICollection<KeyValuePair<string, Type?>>)_backingDictionary).Count;
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs:64]: Why are all these casts of `_backingDictionary` necessary? Can't you just return `_backingDictionary.Count`?

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs (hunk 13 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/XsltOld/Processor.cs:377]: Sure, I can add a unit test for this. This is actually addressing a warning on the old xslt which is not currently covered by this PR (except for this one) but I can add a test here that ensures that using scripts on old xslt throws PNSE at the compile stage which is earlier than this code comes into play.

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs (hunk 14 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/Scripts.cs:64]: I'll fix it, not necessary, I just used the  quick refactoring ctrl+'.' options in VS that added the casts:
image

**Review by eerhardt — APPROVED** [review]
Looks good. Thanks @joperezr!
